src/main/java/seedu/address/model/person/AppointmentByDatePredicate.java

```diff
@@ -0,0 +1,48 @@
+package seedu.address.model.person;
+
+import java.time.LocalDate;
+import java.util.List;
+import java.util.function.Predicate;
+
+/**
+ * Tests that a {@code Person}'s {@code Appointment Date} matches any of the appointment dates given.
+ */
+public class AppointmentByDatePredicate implements Predicate<Person> {
+    private final List<LocalDate> appointments;
+
+    /**
+     * Constructor for AppointmentByDatePredicate.
+     * @param appointments patients' appointments
+     */
+    public AppointmentByDatePredicate(List<LocalDate> appointments) {
+        this.appointments = appointments;
+    }
+
+    /**
+     * Tests if the input argument matches the person's details
+     * @param person the input argument
+     * @return the test result
+     */
+    @Override
+    public boolean test(Person person) {
+        Boolean pastAppointments = appointments.stream().anyMatch(keyword -> person.getPastAppointments().stream()
+                .anyMatch(appointment -> keyword.compareTo(appointment.getDate()) == 0));
+        Boolean upcomingAppointments = appointments.stream().anyMatch(keyword
+                -> person.getUpcomingAppointment().isPresent())
+                && appointments.stream().anyMatch(keyword -> person.getUpcomingAppointment().stream()
+                .anyMatch(appointment -> keyword.compareTo(appointment.getDate()) == 0));
+        return pastAppointments || upcomingAppointments;
+    }
```

Comment:
Great implementation of the `test` method! Perhaps better abstraction for the booleans within `upcomingAppointments` could have been better. For instance, the `upcomingAppointments` variable could be further split into `Boolean hasUpcomingAppt` && `Boolean isSameDate`.